File: scripts/dump-l8-upgrade.mjs
// Dump L8 objective questions whose explanation still carries a stale
// "缺失 / 无法独立核实 / 未经独立核实" claim. Reuses the audit's
// safe-eval approach (preload l8ProgrammingByPaper, mock LEVEL*_TAGS).
import fs from 'fs';
import path from 'path';
import { fileURLToPath } from 'url';
import vm from 'vm';

const __dirname = path.dirname(fileURLToPath(import.meta.url));
const ROOT = path.resolve(__dirname, '..');
const L8 = path.join(ROOT, 'src', 'data', 'gesp', 'level8');
const PROG = path.join(L8, 'programming.js');
const V2 = path.join(ROOT, 'tmp', 'reconstruct-input-v2.json');
const RECON = path.join(ROOT, 'tmp', 'all-reconstructed.json');
const OUT = path.join(ROOT, 'tmp', 'l8-upgrade-input.json');

const recovered = new Set(
  Object.keys(JSON.parse(fs.readFileSync(RECON, 'utf8')))
    .filter(k => k.startsWith('202') && k.split('#')[0].endsWith('l8'))
);
const v2 = JSON.parse(fs.readFileSync(V2, 'utf8'));
const v2map = {};
for (const x of v2) v2map[`${x.paperId}#${x.questionId}`] = x;

const progSrc = fs.readFileSync(PROG, 'utf8');
const l8m = progSrc.match(/const l8ProgrammingByPaper\s*=\s*\{[\s\S]*?\n\};/);
const l8const = l8m ? l8m[0] : 'const l8ProgrammingByPaper = {};';

const optClaimRe = /选项[^。，]*?缺失|选项文本[^。，]*?缺失|选项被污染|选项缺失|选项在源数据中缺失|选项文本在数据中缺失/;
const stemClaimRe = /程序[^。，]*?缺失|代码[^。，]*?缺失|题面[^。，]*?缺失|图示缺失|代码被污染|语句缺失|条件缺失|函数与下界|复杂度符号|题干数字|题干公式|通项公式与求和公式/;
const unverifRe = /无法独立核实|未经独立核实|未经核实|无法核对|不能独立/;
const LETTERS = ['A','B','C','D','E','F'];

const out = [];
for (const fn of fs.readdirSync(L8).filter(f => /^\d{4}-\d{2}-l8\.js$/.test(f))) {
  const fp = path.join(L8, fn);
  let src = fs.readFileSync(fp, 'utf8');
  const paperId = fn.slice(0, -3);
  src = src.replace(/^import .*$/m, '').replace(/^export /gm, '');
  src = src.replace(/const l8ProgrammingByPaper\s*=\s*\{[\s\S]*?\n\};/, '');
  const mock = "const LEVEL8_TAGS = new Proxy({}, { get: () => ({}), getOwnPropertyDescriptor: () => ({}) });";
  const sandbox = {};
  try {
    vm.runInNewContext(`${l8const}\n${mock}\n${src}\nthis.paperData = typeof paperData !== 'undefined' ? paperData : null;`, sandbox, { filename: fn });
  } catch (e) {
    console.error(`EVAL FAIL ${paperId}: ${e.message}`);
    continue;
  }
  const pd = sandbox.paperData;
  if (!pd) { console.error(`NO paperData ${paperId}`); continue; }
  const qs = [...(pd.questions || []), ...(pd.programmingQuestions || [])];
  for (const q of qs) {
    const qid = q.id;
    const key = `${paperId}#${qid}`;
    const expl = q.explanation || '';
    if (!(optClaimRe.test(expl) || stemClaimRe.test(expl) || unverifRe.test(expl))) continue;
    const opts = q.options || [];
    let ansLetter = null;
    if (typeof q.answer === 'number' && q.answer >= 0 && q.answer < opts.length) ansLetter = LETTERS[q.answer];
    else if (typeof q.answer === 'string' && LETTERS.includes(q.answer)) ansLetter = q.answer;
    out.push({
      paperId, questionId: qid, type: q.type,
      options: opts, answer: q.answer, answerLetter: ansLetter,
      stem: (v2map[key] && v2map[key].pdfBody) || '',
      currentExplanation: expl,
      recovered: recovered.has(key),
      optionClaim: optClaimRe.test(expl),
      stemClaim: stemClaimRe.test(expl),
      unverifiedClaim: unverifRe.test(expl),
    });
  }
}

fs.writeFileSync(OUT, JSON.stringify(out, null, 2));
const nRec = out.filter(o => o.recovered).length;
const nStem = out.filter(o => o.stemClaim).length;
const nOpt = out.filter(o => o.optionClaim).length;
console.log(`Total stale L8 questions: ${out.length}`);
console.log(`  recovered-options (47-set): ${nRec}`);
console.log(`  have option-missing claim: ${nOpt}`);
console.log(`  have stem/program-missing claim: ${nStem}`);
console.log(`  recovered AND stem-claim: ${out.filter(o => o.recovered && o.stemClaim).length}`);
console.log(`Wrote ${OUT}`);
for (const o of out) console.log(`  ${o.paperId} Q${o.questionId} type=${o.type} ans=${o.answerLetter} rec=${o.recovered} opt=${o.optionClaim} stem=${o.stemClaim} opts=${o.options.length} stemLen=${o.stem.length}`);

File: scripts/dump-l8-upgrade.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
"""Dump L8 objective questions needing explanation upgrade.
For each L8 question whose explanation still contains a stale
"缺失 / 无法独立核实 / 未经独立核实" claim, extract:
  - type, options (current, recovered), answer letter
  - stem (pdfBody from v2 if available)
  - current explanation
  - flags: recovered (in 47-set), option_claim, stem_claim
Output: tmp/l8-upgrade-input.json
"""
import json, re, os, glob

ROOT = os.path.dirname(os.path.dirname(os.path.abspath(__file__)))
L8 = os.path.join(ROOT, 'src', 'data', 'gesp', 'level8')
PROG = os.path.join(L8, 'programming.js')
V2 = os.path.join(ROOT, 'tmp', 'reconstruct-input-v2.json')
RECON = os.path.join(ROOT, 'tmp', 'all-reconstructed.json')
OUT = os.path.join(ROOT, 'tmp', 'l8-upgrade-input.json')

# 47 recovered qids (options were placeholders, now real)
recon = json.load(open(RECON))
recovered = set(k for k in recon if k.startswith('202') and 'l8' in k.split('#')[0])

# v2 pdfBody by "paperId#qid"
v2 = json.load(open(V2))
v2map = {f"{x['paperId']}#{x['questionId']}": x for x in v2}

# preload l8ProgrammingByPaper so L8 paper files eval cleanly
prog_src = open(PROG, encoding='utf-8').read()
m = re.search(r'const l8ProgrammingByPaper\s*=\s*\{.*?\n\};', prog_src, re.S)
l8const = m.group(0) if m else 'const l8ProgrammingByPaper = {};'

# stale-claim detection
opt_claim_re = re.compile(r'选项[^。，]*?缺失|选项文本[^。，]*?缺失|选项被污染|选项缺失|选项在源数据中缺失|选项文本在数据中缺失')
stem_claim_re = re.compile(r'程序[^。，]*?缺失|代码[^。，]*?缺失|题面[^。，]*?缺失|图示缺失|代码被污染|语句缺失|条件缺失|函数与下界|复杂度符号|题干数字|题干公式|通项公式与求和公式')
unverif_re = re.compile(r'无法独立核实|未经独立核实|未经核实|无法核对|不能独立')

out = []
files = sorted(glob.glob(os.path.join(L8, '*-l8.js')))
for fp in files:
    src = open(fp, encoding='utf-8').read()
    paperId = os.path.basename(fp)[:-3]
    # strip import + export lines; preload l8 const + mock tags
    clean = re.sub(r'^import .*$', '', src, flags=re.M)
    clean = re.sub(r'^export .*$', '', clean, flags=re.M)
    clean = re.sub(r'^const l8ProgrammingByPaper.*?;\s*$', '', clean, flags=re.M|re.S)
    mock = "const LEVEL8_TAGS = new Proxy({}, { get: () => ({}), getOwnPropertyDescriptor: () => ({}) });"
    try:
        ns = {}
        exec(f"{l8const}\n{mock}\n{clean}\n", ns)
    except Exception as e:
        print(f"EVAL FAIL {paperId}: {e}")
        continue
    pd = ns.get('paperData')
    if not pd:
        print(f"NO paperData {paperId}")
        continue
    for q in pd.get('questions', []):
        qid = q.get('id')
        key = f"{paperId}#{qid}"
        expl = q.get('explanation', '') or ''
        if not opt_claim_re.search(expl) and not stem_claim_re.search(expl) and not unverif_re.search(expl):
            continue  # not stale
        opts = q.get('options') or []
        ans = q.get('answer')
        v2e = v2map.get(key, {})
        stem = v2e.get('pdfBody', '')
        rec = key in recovered
        out.append({
            'paperId': paperId,
            'questionId': qid,
            'type': q.get('type'),
            'options': opts,
            'answer': ans,
            'answerLetter': None,
            'stem': stem,
            'currentExplanation': expl,
            'recovered': rec,
            'optionClaim': bool(opt_claim_re.search(expl)),
            'stemClaim': bool(stem_claim_re.search(expl)),
            'unverifiedClaim': bool(unverif_re.search(expl)),
        })

# derive answer letter from options index
letter = ['A','B','C','D','E','F']
for o in out:
    if isinstance(o['answer'], int) and 0 <= o['answer'] < len(o['options']):
        o['answerLetter'] = letter[o['answer']]
    elif isinstance(o['answer'], str) and o['answer'] in letter:
        o['answerLetter'] = o['answer']

json.dump(out, open(OUT, 'w'), ensure_ascii=False, indent=2)
# summary
n_rec = sum(1 for o in out if o['recovered'])
n_stem = sum(1 for o in out if o['stemClaim'])
n_opt = sum(1 for o in out if o['optionClaim'])
print(f"Total stale L8 questions: {len(out)}")
print(f"  recovered-options (47-set): {n_rec}")
print(f"  have option-missing claim: {n_opt}")
print(f"  have stem/program-missing claim: {n_stem}")
print(f"  both: {sum(1 for o in out if o['recovered'] and o['stemClaim'])}")
print(f"Wrote {OUT}")
for o in out:
    print(f"  {o['paperId']} Q{o['questionId']} type={o['type']} ans={o['answerLetter']} rec={o['recovered']} opt={o['optionClaim']} stem={o['stemClaim']} opts={len(o['options'])}")
